### ISSUE TITLE
Add max_retires setting for failed extractions. Look into alternatives for failed-extraction handling.

### ISSUE THREAD

**Issue by hollanbm** [davidnewhall/unpackerr#80]
I've got some archives that have checksum errors. I've tested the rar's and the extracted files are perfectly fine, meaning the checksum error can be ignored.

typically with unrar, this is accomplished using the -kb flag https://linux.die.net/man/1/unrar

I was hoping a config option or env variable would be created, to enable this flag, so that unpacker could successfully unpack the items.

Because the rar's are erroring, unpackerr is continually polling, and attempting to unrar the archives over and over.

I understand this would also add some bad behavior if a file is actually broken. Perhaps unpacker could place these failed archives in a different subfolder on the downloads directory, for manual intervention?

**Comment by davidnewhall** [comment]
How many are like this? Can't you just extract them manually and move on? I've extracted thousands of things and never seen this.

**Comment by hollanbm** [comment]
> How many are like this? Can't you just extract them manually and move on? I've extracted thousands of things and never seen this. EDIT: Please show me the log messages from unpackerr when these files are extracted.

I have about 9 in total. Summing up to around 100 gigs or so. I am currently manually extracting them, and it takes time. I didn't catch this for a week or so. Unpackerr essentially sat there continually extracting and failing over and over. Chewing up resources indefinitely.

`I2020/12/22 14:59:51 Extraction Error: American.Dad.S16E24.720p.WEB.h264-KOGi: rardecode: bad file checksum
2020/12/22 15:03:22 Extraction Error: The.Secret.Life.of.Pets.2016.1080p.BluRay.x264-SPARKS: rardecode: bad file checksum
2020/12/22 15:12:34 Extraction Error: Identity.Thief.2013.UNRATED.1080p.BluRay.x264-SPARKS: rardecode: bad file checksum
2020/12/22 15:15:41 Extraction Error: Justice.League.vs.the.Fatal.Five.2019.1080p.BluRay.x264-SPRiNTER: rardecode: bad file checksum
2020/12/22 15:28:26 Extraction Error: Life.As.We.Know.It.2010.1080p.BluRay.x264-SECTOR7: rardecode: bad file checksum
2020/12/22 15:35:02 Extraction Error: Mary.Poppins.1964.1080p.BluRay.X264-AMIABLE: rardecode: bad file checksum
2020/12/22 15:39:29 Extraction Error: Free.Solo.2018.LiMiTED.1080p.BluRay.x264-CADAVER: rardecode: bad file checksum
2020/12/22 15:45:19 Extraction Error: Wall.Street.Money.Never.Sleeps.2010.1080p.BluRay.x264-Japhson: rardecode: bad file checksum
2020/12/22 15:51:18 Extraction Error: RoboCop.2.1990.1080p.BluRay.x264-LEVERAGE: rardecode: bad file checksum
2020/12/22 15:57:42 Extraction Error: The.Manchurian.Candidate.1962.1080p.BluRay.X264-AMIABLE: rardecode: bad file checksum
2020/12/22 15:57:47 Extraction Error: The.Call.2013.1080p.BluRay.x264-SPARKS: rardecode: bad file checksum
`

Adding -kb and extracting to a temp directory would have saved some time. Another option would be to disable the history reset. Currently I can set a time on how long before an items history is cleared, and it is treated as new. But I don't have the ability to completely disable that functionality, just set an extremely large number. Which also doesn't seem optimal.

**Comment by davidnewhall** [comment]
Thank you for the logs!

What comes to mind is that something is corrupting your files. Either a downloader, or an intermediate process. You should figure that out. I can almost guarantee that the files you extract will have glitches, pops and problems in general when playing them back; hopefully it's insignificant and toward the beginning or end.

As for unpackerr. It doesn't use `unrar`. It _is_ unrar. I'd have to fabricate (write code for) the behavior you're expecting, and it's not entirely clear what's appropriate here. I think 99% of users would rather this fail so they can go download the content again, undamaged. 

Do you want a flag to disable extraction errors entirely? Only items that get marked error will reset and start over, so if I remove that designation [with a config option] it would just do the best it can and let the item sit in `extracted` forever. Or, we can add a config option to not reset items that failed. 

I cannot make any promises that this tool will ever extract damaged files the same way `unrar -kb` does; I'm just not sure I have the ability to continue extracting once we hit errors, but if you think there's value in that I can give it a try. Reproducing the problem may not be very easy for me though.

**Comment by hollanbm** [comment]
OHHHH, I assumed you were using the unrar packages to process extractions. Disregard the -kb flag then, I figured that would just be a quick fix.

In this scenario, in a perfect world, I'd like to have unpackerr stop after X attempts, and not clear the history

I will investigate and see what was causing the corruption. I did a cursory scan through the files and all seemed fine. These files are stored on ZFS, and accessed via NFS over a 10gb link. Any ideas on where to start with the corruption? I had done some rechecks in the client to make sure, and they passed at 100%.

Any ideas on where to start investigating corruption? ubuntu 20.04.1 host

**Comment by davidnewhall** [comment]
I can look into adding a `max_retries` setting. That seems quite reasonable and useful.

**Comment by davidnewhall** [comment]
Added a maximum retries. I'll do what I can to actually test that it works, but I think it should be good. I sure hope you found the source of corrupt files. If not, maybe try copying a folder without NFS and see if the corruption persists. I also wonder if you can run unpackerr on the server that has the files. That'd be considerably more reliable than extracting over NFS.

**Comment by hollanbm** [comment]
Yes, I was able to track it down. It was my new client, qBittorrent. I found quite a few threads and posts documenting similar behavior. Where a torrent would complete, then become corrupt. I'm switching to transmission, and hopefully I'll have better success.

What led me to qBT was the speeds I was getting. I was saturating my WAN connection. I'm guessing qBT takes some liberties with how blocks/chunks are handled, which explains the speed gains, and also by corruption issue.

I used to use deluge, but I got frustrated with some of its bugs, and wanted to try something new. Glad I did cause it led me to unpackerr (which is much more elegant than the 1 line find command I had the clients calling on completion)

The rest of my setup is dockerized. I'm making use of docker volumes for my NFS mounts, I've also got sync writes enabled on the backend. The containers themselves never know it's an NFS volume, and with 10gbe the bottleneck is my array speed.
